File: erp-backend/src/main/java/com/erp/controller/OrderController.java
package com.erp.controller;

import com.erp.dto.BookOrderRequest;
import com.erp.dto.ReturnOrderRequest;
import com.erp.entity.*;
import com.erp.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
public class OrderController {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private OrderItemRepository orderItemRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private LedgerRepository ledgerRepository;

    @Autowired
    private ShopRepository shopRepository;

    // POST: Multiple items ke saath order book karna
    @PostMapping("/book-order")
    @Transactional
    public ResponseEntity<?> bookOrder(
            @RequestAttribute("role") String role,
            @RequestBody BookOrderRequest request) {
        // ✅ SECURITY FIX: Only admin, staff, or shopkeeper can book orders
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role) && !"shopkeeper".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body(Map.of("message", "Access Denied: Unauthorized role"));
        }

        // A. Orders table mein main entry
        Order order = new Order();
        order.setShopId(request.getShop_id());
        order.setUserId(request.getUser_id());
        order.setTotalAmount(request.getTotal_amount());
        order.setStatus("pending");
        Order savedOrder = orderRepository.save(order);

        // B. Order Items table mein har item ki entry
        for (BookOrderRequest.OrderItemRequest item : request.getItems()) {
            OrderItem orderItem = new OrderItem();
            orderItem.setOrderId(savedOrder.getOrderId());
            orderItem.setProductId(item.getProduct_id());
            orderItem.setQuantity(item.getQuantity());
            orderItem.setPriceAtSale(item.getPrice());
            orderItemRepository.save(orderItem);

            // C. Automatically update stock
            productRepository.reduceStock(item.getProduct_id(), item.getQuantity());
        }

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("message", "Order booked with items!");
        response.put("order", savedOrder);
        return ResponseEntity.ok(response);
    }

    // PUT: Order Deliver karna aur Ledger Update karna
    @PutMapping("/deliver-order/{order_id}")
    @Transactional
    public ResponseEntity<?> deliverOrder(
            @RequestAttribute("role") String role,
            @PathVariable Integer order_id) {
        // ✅ SECURITY FIX: Only admin or staff can mark orders as delivered
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin or Staff can deliver orders");
        }

        // 1. Order ka data nikalna
        Order order = orderRepository.findById(order_id).orElseThrow();
        Integer shopId = order.getShopId();
        BigDecimal totalAmount = order.getTotalAmount();

        // 2. Status 'delivered' karna
        order.setStatus("delivered");
        orderRepository.save(order);

        // 3. Ledger mein "Udhaar" add karna
        List<Ledger> lastEntries = ledgerRepository.findLatestByShopId(shopId);
        BigDecimal oldBalance = lastEntries.isEmpty() ? BigDecimal.ZERO : lastEntries.get(0).getBalance();
        BigDecimal newBalance = oldBalance.add(totalAmount);

        Ledger ledger = new Ledger();
        ledger.setShopId(shopId);
        ledger.setDescription("Order Delivered (ID: " + order_id + ")");
        ledger.setDebit(totalAmount);
        ledger.setCredit(BigDecimal.ZERO);
        ledger.setBalance(newBalance);
        ledgerRepository.save(ledger);

        // 4. Shop total_debt update
        Shop shop = shopRepository.findById(shopId).orElseThrow();
        shop.setTotalDebt(newBalance);
        shopRepository.save(shop);

        return ResponseEntity.ok("Order delivered and Ledger updated!");
    }

    // GET: Saare pending orders
    @GetMapping("/pending-orders")
    public ResponseEntity<?> getPendingOrders(@RequestAttribute("role") String role) {
        // ✅ SECURITY FIX: Only admin or staff can view all pending orders
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin or Staff can view pending orders");
        }
        return ResponseEntity.ok(orderRepository.findByStatus("pending"));
    }

    // GET: Get orders list, optionally filtered by status
    @GetMapping("/orders")
    public ResponseEntity<?> getOrders(
            @RequestAttribute("role") String role,
            @RequestParam(required = false) String status) {
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role) && !"shopkeeper".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied");
        }
        if (status == null || status.isBlank()) {
            return ResponseEntity.ok(orderRepository.findAll(org.springframework.data.domain.Sort.by(org.springframework.data.domain.Sort.Direction.DESC, "orderId")));
        }
        return ResponseEntity.ok(orderRepository.findByStatus(status));
    }

    // GET: Get all items inside an order with product names
    @GetMapping("/order-items/{orderId}")
    public ResponseEntity<?> getOrderItems(
            @RequestAttribute("role") String role,
            @PathVariable Integer orderId) {
        if (role == null || role.isBlank()) {
            return ResponseEntity.status(403).body("Access Denied");
        }
        
        List<OrderItem> items = orderItemRepository.findByOrderId(orderId);
        List<Map<String, Object>> result = new java.util.ArrayList<>();
        
        for (OrderItem item : items) {
            Map<String, Object> map = new java.util.LinkedHashMap<>();
            map.put("itemId", item.getItemId());
            map.put("productId", item.getProductId());
            map.put("quantity", item.getQuantity());
            map.put("priceAtSale", item.getPriceAtSale());
            map.put("returnedQuantity", item.getReturnedQuantity() != null ? item.getReturnedQuantity() : 0);
            
            if (item.getProductId() != null) {
                Product p = productRepository.findById(item.getProductId()).orElse(null);
                if (p != null) {
                    map.put("itemName", p.getItemName());
                    map.put("brandName", p.getBrandName());
                } else {
                    map.put("itemName", "Unknown Product");
                    map.put("brandName", "N/A");
                }
            } else {
                map.put("itemName", "Deleted Product");
                map.put("brandName", "N/A");
            }
            result.add(map);
        }
        
        return ResponseEntity.ok(result);
    }

    // PUT: Order Cancel karna aur Stock Restore karna
    @PutMapping("/cancel-order/{order_id}")
    @Transactional
    public ResponseEntity<?> cancelOrder(
            @RequestAttribute("role") String role,
            @PathVariable Integer order_id) {
        // ✅ SECURITY FIX: Only admin or staff can cancel orders
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin or Staff can cancel orders");
        }

        Order order = orderRepository.findById(order_id).orElseThrow();
        if (!"pending".equalsIgnoreCase(order.getStatus())) {
            return ResponseEntity.badRequest().body("Only pending orders can be cancelled!");
        }

        // 1. Status 'cancelled' karna
        order.setStatus("cancelled");
        orderRepository.save(order);

        // 2. Product stock restore karna
        List<OrderItem> items = orderItemRepository.findByOrderId(order_id);
        for (OrderItem item : items) {
            if (item.getProductId() != null) {
                productRepository.restoreStock(item.getProductId(), item.getQuantity());
            }
        }

        return ResponseEntity.ok("Order cancelled and stock restored!");
    }

    // POST: Return Items from a delivered order
    @PostMapping("/return-order/{orderId}")
    @Transactional
    public ResponseEntity<?> returnOrder(
            @RequestAttribute("role") String role,
            @PathVariable Integer orderId,
            @RequestBody ReturnOrderRequest request) {
        // Only admin or staff can return orders
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin or Staff can process returns");
        }

        Order order = orderRepository.findById(orderId).orElse(null);
        if (order == null) {
            return ResponseEntity.status(404).body("Order not found");
        }

        // Only delivered or partially returned orders can have returns
        if (!"delivered".equalsIgnoreCase(order.getStatus()) && !"partially returned".equalsIgnoreCase(order.getStatus())) {
            return ResponseEntity.badRequest().body("Only delivered or partially returned orders can be returned!");
        }

        List<OrderItem> orderItems = orderItemRepository.findByOrderId(orderId);
        BigDecimal totalRefund = BigDecimal.ZERO;

        for (ReturnOrderRequest.ReturnItemRequest returnItem : request.getItems()) {
            Integer prodId = returnItem.getProductId();
            Integer retQty = returnItem.getQuantity();

            if (retQty == null || retQty <= 0) {
                continue;
            }

            // Find matching OrderItem
            OrderItem matchedItem = null;
            for (OrderItem oi : orderItems) {
                if (oi.getProductId() != null && oi.getProductId().equals(prodId)) {
                    matchedItem = oi;
                    break;
                }
            }

            if (matchedItem == null) {
                return ResponseEntity.badRequest().body("Product ID " + prodId + " is not part of this order!");
            }

            int currentReturned = matchedItem.getReturnedQuantity() != null ? matchedItem.getReturnedQuantity() : 0;
            if (currentReturned + retQty > matchedItem.getQuantity()) {
                return ResponseEntity.badRequest().body("Return quantity " + retQty + " exceeds remaining returnable quantity for product ID " + prodId);
            }

            // 1. Update returned quantity on OrderItem
            matchedItem.setReturnedQuantity(currentReturned + retQty);
            orderItemRepository.save(matchedItem);

            // 2. Restore stock for product
            productRepository.restoreStock(prodId, retQty);

            // 3. Accumulate refund amount
            BigDecimal itemRefund = matchedItem.getPriceAtSale().multiply(new BigDecimal(retQty));
            totalRefund = totalRefund.add(itemRefund);
        }

        if (totalRefund.compareTo(BigDecimal.ZERO) > 0) {
            // 4. Determine new Order status
            boolean allReturned = true;
            for (OrderItem oi : orderItems) {
                int returned = oi.getReturnedQuantity() != null ? oi.getReturnedQuantity() : 0;
                if (returned < oi.getQuantity()) {
                    allReturned = false;
                    break;
                }
            }
            order.setStatus(allReturned ? "returned" : "partially returned");
            orderRepository.save(order);

            // 5. Update Ledger with Credit
            Integer shopId = order.getShopId();
            List<Ledger> lastEntries = ledgerRepository.findLatestByShopId(shopId);
            BigDecimal oldBalance = lastEntries.isEmpty() ? BigDecimal.ZERO : lastEntries.get(0).getBalance();
            BigDecimal newBalance = oldBalance.subtract(totalRefund);

            Ledger ledger = new Ledger();
            ledger.setShopId(shopId);
            ledger.setDescription("Returned Items (Order ID: " + orderId + ")");
            ledger.setDebit(BigDecimal.ZERO);
            ledger.setCredit(totalRefund);
            ledger.setBalance(newBalance);
            ledgerRepository.save(ledger);

            // 6. Update Shop total debt
            Shop shop = shopRepository.findById(shopId).orElseThrow();
            shop.setTotalDebt(newBalance);
            shopRepository.save(shop);
        } else {
            return ResponseEntity.badRequest().body("No valid items were returned!");
        }

        return ResponseEntity.ok(Map.of(
            "message", "Return processed successfully",
            "refundAmount", totalRefund,
            "status", order.getStatus()
        ));
    }
}

File: erp-backend-node/src/controllers/orderController.ts
import { Request, Response } from 'express';
import prisma from '../prisma/client';

// Book a new order
export const bookOrder = async (req: Request, res: Response) => {
  try {
    const { shop_id, items, total_amount, user_id } = req.body;
    
    const parsedShopId = shop_id ? parseInt(shop_id) : null;
    const parsedUserId = user_id ? parseInt(user_id) : null;
    const parsedTotal = Number(total_amount || 0);

    // Use a transaction to ensure everything saves together
    const result = await prisma.$transaction(async (tx) => {
      // 1. Create the order
      const newOrder = await tx.orders.create({
        data: {
          shop_id: parsedShopId,
          total_amount: parsedTotal,
          user_id: parsedUserId,
          status: 'pending',
          order_date: new Date()
        }
      });

      // 2. Add order items and reduce stock
      for (const item of items) {
        const prodId = parseInt(item.product_id || item.productId);
        const qty = Number(item.quantity);
        const price = Number(item.price_at_sale || item.priceAtSale);

        // Create order item
        await tx.order_items.create({
          data: {
            order_id: newOrder.order_id,
            product_id: prodId,
            quantity: qty,
            price_at_sale: price,
            returned_quantity: 0
          }
        });

        // Reduce product stock
        await tx.products.update({
          where: { id: prodId },
          data: {
            stock: { decrement: qty }
          }
        });
      }

      // 3. Add to shop debt and ledger (Assuming total_amount is added as debt)
      await tx.shops.update({
        where: { shop_id },
        data: {
          total_debt: { increment: total_amount }
        }
      });

      // Add a ledger entry (Debit)
      const shop = await tx.shops.findUnique({ where: { shop_id } });
      await tx.ledger.create({
        data: {
          shop_id,
          description: `Order #${newOrder.order_id}`,
          debit: total_amount,
          credit: 0,
          balance: shop?.total_debt,
          date: new Date()
        }
      });

      return newOrder;
    });

    res.json({ message: 'Order booked successfully', orderId: result.order_id });
  } catch (error) {
    console.error('Error booking order:', error);
    res.status(500).json({ message: 'Server error while booking order.' });
  }
};

// Get pending orders
export const getPendingOrders = async (req: Request, res: Response) => {
  try {
    const orders = await prisma.orders.findMany({
      where: { status: 'pending' },
      include: {
        shops: true,
        users: true
      },
      orderBy: { order_date: 'desc' }
    });
    res.json(orders);
  } catch (error) {
    console.error('Error fetching pending orders:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};

// Get all orders
export const getOrders = async (req: Request, res: Response) => {
  try {
    const orders = await prisma.orders.findMany({
      include: {
        shops: true,
        users: true
      },
      orderBy: { order_date: 'desc' }
    });
    res.json(orders);
  } catch (error) {
    console.error('Error fetching orders:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};

// Get order items by order ID
export const getOrderItems = async (req: Request, res: Response) => {
  try {
    const orderId = req.params.orderId as string;
    const items = await prisma.order_items.findMany({
      where: { order_id: parseInt(orderId) },
      include: { products: true }
    });
    res.json(items);
  } catch (error) {
    console.error('Error fetching order items:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};

// Deliver an order
export const deliverOrder = async (req: Request, res: Response) => {
  try {
    const orderId = parseInt(req.params.orderId as string);
    if (isNaN(orderId)) {
      return res.status(400).json({ message: 'Invalid order ID format. Please refresh your browser.' });
    }

    await prisma.orders.update({
      where: { order_id: orderId },
      data: { status: 'delivered' }
    });
    res.json({ message: 'Order marked as delivered.' });
  } catch (error) {
    console.error('Error delivering order:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};

// Cancel an order (Revert stock and ledger)
export const cancelOrder = async (req: Request, res: Response) => {
  try {
    const orderId = parseInt(req.params.orderId as string);
    if (isNaN(orderId)) {
      return res.status(400).json({ message: 'Invalid order ID format. Please refresh your browser.' });
    }
    
    await prisma.$transaction(async (tx) => {
      const order = await tx.orders.findUnique({ 
        where: { order_id: orderId },
        include: { order_items: true } 
      });

      if (!order || order.status === 'cancelled') return;

      // Revert stock
      for (const item of order.order_items) {
        await tx.products.update({
          where: { id: item.product_id! },
          data: { stock: { increment: item.quantity } }
        });
      }

      // Revert debt
      if (order.shop_id && order.total_amount) {
        await tx.shops.update({
          where: { shop_id: order.shop_id },
          data: { total_debt: { decrement: order.total_amount } }
        });

        // Add ledger entry for cancellation (Credit)
        const shop = await tx.shops.findUnique({ where: { shop_id: order.shop_id } });
        await tx.ledger.create({
          data: {
            shop_id: order.shop_id,
            description: `Order #${order.order_id} Cancelled`,
            debit: 0,
            credit: order.total_amount,
            balance: shop?.total_debt,
            date: new Date()
          }
        });
      }

      // Mark order as cancelled
      await tx.orders.update({
        where: { order_id: orderId },
        data: { status: 'cancelled' }
      });
    });

    res.json({ message: 'Order cancelled successfully.' });
  } catch (error) {
    console.error('Error cancelling order:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};

// Return order items
export const returnOrder = async (req: Request, res: Response) => {
  try {
    const orderId = parseInt(req.params.orderId as string);
    if (isNaN(orderId)) {
      return res.status(400).json({ message: 'Invalid order ID format. Please refresh your browser.' });
    }

    const { items } = req.body;
    
    await prisma.$transaction(async (tx) => {
      const order = await tx.orders.findUnique({ where: { order_id: orderId } });
      if (!order) throw new Error("Order not found");

      let totalRefund = 0;

      for (const item of items) {
        const prodId = parseInt(item.productId || item.product_id);
        const qty = Number(item.quantity);

        // Find original order item
        const orderItem = await tx.order_items.findFirst({
          where: { order_id: orderId, product_id: prodId }
        });

        if (orderItem) {
          totalRefund += Number(orderItem.price_at_sale) * qty;
          
          await tx.order_items.update({
            where: { item_id: orderItem.item_id },
            data: { returned_quantity: { increment: qty } }
          });

          await tx.products.update({
            where: { id: prodId },
            data: { stock: { increment: qty } }
          });
        }
      }

      // Deduct from debt
      if (order.shop_id && totalRefund > 0) {
        await tx.shops.update({
          where: { shop_id: order.shop_id },
          data: { total_debt: { decrement: totalRefund } }
        });

        const shop = await tx.shops.findUnique({ where: { shop_id: order.shop_id } });
        await tx.ledger.create({
          data: {
            shop_id: order.shop_id,
            description: `Returned items for Order #${orderId}`,
            debit: 0,
            credit: totalRefund,
            balance: shop?.total_debt,
            date: new Date()
          }
        });
      }

      // Mark order as partially returned or returned
      await tx.orders.update({
        where: { order_id: orderId },
        data: { status: 'partially returned' }
      });
    });

    res.json({ message: 'Return processed successfully.' });
  } catch (error) {
    console.error('Error returning order:', error);
    res.status(500).json({ message: 'Server error.' });
  }
};
